src/MOCKernel.cpp

```diff
@@ -0,0 +1,165 @@
+ *        data from
+ * @param row_num the row index into the temporary segments matrix
+ */
+CounterKernel::CounterKernel(TrackGenerator* track_generator, int row_num) :
+                           MOCKernel(track_generator, row_num) {}
+
+
+/**
+ * @brief Prepares an MOCKernel for a new Track
+ * @details Resets the segment count
+ * @param track The new Track the MOCKernel prepares to handle
+ */
+void MOCKernel::newTrack(Track* track) {
+  _count = 0;
+}
+
+
+/**
+ * @brief Prepares a VolumeKernel for a new Track
+ * @details Resets the segment count and updates the weight for the new Track
+ * @param track The new Track the MOCKernel prepares to handle
+ */
+void VolumeKernel::newTrack(Track* track) {
+
+  /* Compute the Track cross-sectional area */
+  int azim_index = track->getAzimAngleIndex();
+  _weight = _quadrature->getAzimSpacing(azim_index)
+      * _quadrature->getAzimWeight(azim_index);
+
+  /* Reset the count */
+  _count = 0;
+}
+
+
+/**
+ * @brief Destructor for MOCKernel
+ */
+MOCKernel::~MOCKernel() {};
+
+
+/*
+ * @brief Reads and returns the current count
+ * @details MOC kernels count how many times they are accessed. This value
+ *          returns the value of the counter (number of execute accesses)
+ *          since kernel creation or last call to newTrack.
+ * @return _count the counter value
+ */
+int MOCKernel::getCount() {
+  return _count;
+}
+
+
+/* @brief Resets the maximum optcal path length for a segment
+ * @details MOC kernels ensure that there are no segments with an optical path
+ *          length greater than the maximum optical path length by splitting
+ *          then when they get too large.
+ * @param the maximum optical path length for a segment
+ */
+void MOCKernel::setMaxOpticalLength(FP_PRECISION max_tau) {
+  _max_tau = max_tau;
+}
+
+
+/*
+ * @brief Adds segment contribution to the FSR volume
+ * @details The VolumeKernel execute function adds the product of the
+ *          track length and track weight to the buffer array at index
+ *          id, referring to the array of FSR volumes.
+ * @param length segment length
+ * @param mat Material associated with the segment
+ * @param id the FSR ID of the FSR associated with the segment
+ */
+void VolumeKernel::execute(FP_PRECISION length, Material* mat, int id,
+    int cmfd_surface_fwd, int cmfd_surface_bwd) {
+
+  /* Set omp lock for FSRs */
+  omp_set_lock(&_FSR_locks[id]);
+
+  /* Add value to buffer */
+  _FSR_volumes[id] += _weight * length;
+
+  /* Unset lock */
+  omp_unset_lock(&_FSR_locks[id]);
+
+  /* Determine the number of cuts on the segment */
+  FP_PRECISION* sigma_t = mat->getSigmaT();
+  double max_sigma_t = 0;
+  for (int e=0; e < mat->getNumEnergyGroups(); e++)
+    if (sigma_t[e] > max_sigma_t)
+      max_sigma_t = sigma_t[e];
+
+  int num_cuts = std::max((int) std::ceil(length * max_sigma_t / _max_tau), 1);
+
+  /* Increment count */
+  _count += num_cuts;
```

Comment:
It looks like `_count` is used for various different things in the different kernels? Above you define it in a comment as the number of times `execute` is accessed, but here it is obviously not that and is instead the number of segment cuts.